bin/runtime-common/src/messages.rs

```diff
@@ -323,33 +323,21 @@ pub mod target {
 		MessageNonce,
 	);
 
-	/// Message payload for Bridged -> This messages.
-	pub struct FromBridgedChainMessagePayload<B: MessageBridge>(pub(crate) FromBridgedChainDecodedMessagePayload<B>);
-
-	impl<B: MessageBridge> From<FromBridgedChainDecodedMessagePayload<B>> for FromBridgedChainMessagePayload<B> {
-		fn from(decoded_payload: FromBridgedChainDecodedMessagePayload<B>) -> Self {
-			Self(decoded_payload)
-		}
+	/// Encoded Call of This chain as it is transferred over bridge.
+	///
+	/// Our Call is opaque (`Vec<u8>`) for Bridged chain. So it is encoded, prefixed with
+	/// vector length. Custom decode implementation here is exactly to deal with this.
+	#[derive(Decode, Encode, RuntimeDebug, PartialEq)]
+	pub struct FromBridgedChainEncodedMessageCall<B> {
```

Comment:
I'd ditch the this type here and `type EncodedCall` and rather move that logic to `call-dispatch`, like so:

```rust
#[derive(RuntimeDebug, Encode, Decode, Clone, PartialEq, Eq)]
pub struct MessagePayload<SourceChainAccountId, TargetChainAccountPublic, TargetChainSignature, Call> {
	/// Runtime specification version. We only dispatch messages that have the same
	/// runtime version. Otherwise we risk to misinterpret encoded calls.
	pub spec_version: SpecVersion,
	/// Weight of the call, declared by the message sender. If it is less than actual
	/// static weight, the call is not dispatched.
	pub weight: Weight,
	/// Call origin to be used during dispatch.
	pub origin: CallOrigin<SourceChainAccountId, TargetChainAccountPublic, TargetChainSignature>,
	/// The call itself.
	pub call: OpaqueCall<Call>,
}

#[derive(Decode, Encode)]
pub struct OpaqueCall<T>(Vec<u8>, #[codec(skip)] PhantomData<T>);

impl OpaqueCall<T: Decode> {
  pub fn decode_inner(&self) -> Result<...> {
     T::decode(&mut &*self.0)
  }
}
```

IMHO this should also be coupled with #664, so that we don't have extra `note_rejected_message`  function, but rather let the `call-dispatch` do the decoding and spec-version checking and then emit one of 4 events:
1. Spec version failed
2. Decoding Failed
3. Dispatch Failed
4. Dispatch Successful

@svyatonik what do you think?

Reply:
I think that:
1) any parts of delivery logic (and the fact that call is 'strangely encoded' is a part of that logic) shall not be hardcoded in dispatch module. Today the call is `call.encode().encode()`, tomorrow (if we'll deal with offline storage), it'll be simply `call.encode()`;
2) if we don't have an access to the `Call` from outside (and it is what you're suggesting here), then we'll be unable to perform any filtering. Like initially iirc we were going to only dispatch balance calls;
3) you're confusing `Call` decoding (this PR) with message decoding (#614). We need to decode message before dispatch to be able to verify that the relayer has reserved enough dispatch weight for all messages.

Reply:
1. IMHO The call dispatch should be the only module responsible for intepreting the payload of the message. The fact that it has to be double encoded is simply a limitation of SCALE codec we use. Even if we have offline storage, I guess it will look the same if we want to keep dispatch and message decoding separate (which I think we should do).
2. Perhaps that signifies that the filtering should be done in the `call-dispatch` pallet then? I know it complicates things a bit, but might be worth it. The `Call` type is actually known in the delivery pallet, but it's not decoded yet when the filtering is performed, we could pass the filter closure to the `call-dispatch::dispatch()` call so it's executed after we successfully decode the call.
3. No, I think we are on the same page. Message decoding should happen first - if the message can't be decoded we should simply fail the entire transaction. Decoded message will have `weight` parameter so that we can compute the total cost. Next the delivery pallet triggers dispatch (generic) for every message and that's where the decoding & spec_version checking should happen. I.e. delivery (and decoding) of messages should be successful despite `spec_version` mismatch, it's just the dispatch that should fail.

Reply:
1. `send_message` call of the message lane pallet accepts `Vec<u8>`, which is `call.encode()`. I agree that this first decoding should be made only by dispatch pallet. But it is not dispatch pallet who is adding that second `.encode()`. And it should not deal with that decode imo. If we'll migrate to offline storage. I didn't get your argument about keeping separating dispatch + decoding. Isn't separation is what you're suggesting to remove here (by decoding in dispatch)?
2. I think filtering should be done in runtime. Either by providing closure to the dispatch, or directly, if runtime would have decoded `Call`. I'm just saying that in your scenario `Call` is completely hidden from outside world && this won't work.
3. We can't fail entire transaction, because it'll stuck lane. We have just received proof that the other chain has sent this message (which is encoded invalid) to us - why we should fail transaction?

Reply:
Okay, thanks for clarifications, I see your points better now.
1. I'm opting for delivery doing `Vec<u8> -> OpaqueCall` conversion (which would be `Vec::<u8>::encode()`) on `send_message` and then symmetrically delivery pallet doing `OpaqueCall -> Vec<u8>` decoding before handing over to the dispatch pallet. I.e. to overcome the scale issue of just lazily decoding raw bytes.
2. Yes, totally agree it's in the runtime. AFAIR currently we implement `LaneMessageVerifier` and it's supposed to get decoded payload (call). I think we should keep that API, but with `call-dispatch` responsible for decoding (`Vec<u8> -> Call`) we can't do filtering at this level.
Instead I'd propose to add `BaseCallFilter`-like mechanism (this will allow us to re-use existing API) to `call-dispatch` pallet.
3. I think I'm conflating a bunch of things when saying "message" and "payload", hence confusion. But I think I see your point now, we need to attempt to decode the call to get the exact `weight` before we even start doing the dispatch. But I still think it's fine with `OpaqueCall` approach - I'd just prefer to avoid adding generic types, where we can simply add concrete types and make the process more explicit in the pallet. Similarly using `Into` for heavy stuff like `decoding` and getting a result is imho not the right way. We can be explicit that we don't `.into()` between formats, but we simply `.decode()` to overcome SCALE limitations.

I can maybe try to implement a pseudo-code of what I have in mind, perhaps there is something that's obvious to you and what I'm missing which prevents us to get to shared understanding.

Reply:
1 + 3. Ok, so what I got from these points,  is that all this comes from preference of explicit `Decode` + `Vec<u8>` over `EncodedCall`. I have two things to say about this:
1.1) I still do not think that `OpaqueCall` belongs to the `call-dispatch` pallet. Imo, it is about encoding + decoding and shall not be a part of this pallet => should be solved outside (on "transport" level). I haven't seen any strong arguments to not to keep it outside, other than personal preferences;
1.2) I agree that `Into` is awkward, as I already stated in description;
2. That is wrong - `LaneMessageVerifier` works on the source side. On target chain there's only call dispatch, which supposed to do filtering. Basically - before this PR you had `Call` [here](https://github.com/paritytech/parity-bridges-common/pull/663/files#diff-ded5af982e34b56ffecfe2ffeba389edefd5dc7302c45a6c1853361a7e6d749bL389) and could have just drop the `Call` there, before it reache `dispatch` method. Now you can do the same from `Into<Result<...>>`, which is I agree, is also a bit awkward.
3. So given that you agree that the `MessagePayload` should be decoded before dispatch method, does it mean that [this comment](https://github.com/paritytech/parity-bridges-common/pull/664#pullrequestreview-572186156) is now invalid?

What I can suggest is to change `EncodedCall: Into<Result<...>>;` to something like that:
```rust
enum CallError { Decode, Filter }

trait EncodedCall<Call> {
  fn decode_and_filter_call(self) -> Result<Call, DecodeOrFilterError>;
}
```

So what it changes is simply that your `OpaqueCall` would live outside of the pallet (it already lives there as `FromBridgedChainEncodedMessageCall`). The main change is removal of that `Into` conversion.

I'd even go further and introduce more traits, like:
```rust
trait EncodedMessage<Message> {
	/// Resolves #664 - fails if message decode has failed and we can't even verify spec_version, ...
	/// The pallet would emit `MessageRejected` event.
	fn decode(self) -> Result<Message, ()>;
}

trait EncodedCall<Call> {
	/// Fails if call can not be decoded.
	/// The pallet would emit `MessageCallDecodeFailed` event.
	fn decode(self) -> Result<Call, ()>;
}

trait CallFilter<Call> {
	/// Fails if call shall not be dispatched.
	/// The pallet would emit `CallRejected` event.
	fn filter(self) -> Result<Call, ()>;
}
```

But I assume that'll be too much.

Reply:
> I still do not think that `OpaqueCall` belongs to the call-dispatch pallet.

Agree, that was my initial proposal, cause I thought it would be the `call-dispatch` doing the decoding, it should be separate, but it shouldn't be runtime-specific either.

> So given that you agree that the MessagePayload should be decoded before dispatch method, does it mean that this comment is now invalid?

Yeah, let's continue there.

> But I assume that'll be too much.

I'd aim for introducing 0 new traits :)
1. It seems to me that `trait EncodedMessage<Mesage>` is simply `parity_scale_codec::Decode`? 
```rust
// in message delivery
T::Dispatch::dec
2. `CallFilter` is already in `frame_support::traits::Filter` and it has a bunch of useful tools to construct the filters: https://github.com/paritytech/substrate/blob/15fc239ed63c902a225cfb193ca1646a45d34931/frame/support/src/traits.rs#L44


Anyways, I feel we are going nowhere, because I'm basing most of my suggestions on gut feeling that there is too much abstraction, but in reality it's probably not possible to implement them in my suggested form without changing too much of other existing abstractions.

Let's just proceed with explicit decoding (no `Into`) and we can revist later on.

Reply:
> It seems to me that trait EncodedMessage<Mesage> is simply parity_scale_codec::Decode?
> Let's just proceed with explicit decoding (no Into) and we can revist later on.

I think we have lost the track. The idea is that call-dispatch shall not know how the call is transmitted - it may be encoded with parity-codec, it may be `call.encode().encode()`, it may use custom encoding, or it is crafted manually at the target chain, ... This pallet imo shall just dispatch the call and emit events. Like when we have had conversation about Polkadot <-> Kusama bridge, one of ideas was to use `enum Message { TransferBalance(Source, Target) }`, or message could be just `u64`, which is a balance that must be transferred from some hardcoded account A to account B.

I'm not sure I got your idea about replacing `EncodedMessage::decode() -> Call` with `Decode::decode(&mut &input) -> Call`. Where we're going to get this `input`? It seems to me that you're just suggesting exactly what you have been suggesting initially - i.e. to change type of `pallet_bridge_call_dispatch::MessagePayload::call` to `Vec<u8>` and then just manually call `Call::decode()` on it. That's exactly what I'm stand against in this whole discussion.

> CallFilter is already in frame_support::traits::Filter

Ok, then we may reuse existing trait here.

Reply:
Let's proceed with it as-is, I initially thought it could be simplified significantly, but there are caveats that I didn't take into account (like getting the weight first).

While I don't agree with some of your reasoning for increased flexibility (i.e. supporting custom encodings, etc) it would be dumb for me to be strong on reverting the functionality if it's already here.